### ISSUE TITLE
showExpanded option not working on Projects.all

### ISSUE THREAD

**Issue by nivo33** [jdalrymple/gitbeaker#1897]
**Description**

on `Projects.all` method, adding `showExpanded: true` does not show the expanded results as stated in the docs, meaning that there is no way to track if all the results have been returned.

**Steps to reproduce**

Heres my code:
```
const projects = new Projects({token: xxxxx})

const resp = await projects.all({
      showExpanded: true,
      perPage: 20,
      maxPages: 2,
      minAccessLevel: 40,
    })

console.log(resp.data) // undefined
console.log(resp.paginationInfo) // undefined
console.log(resp) // array of repos 
```

There are 98 repos in total if that helps.

**Expected behaviour**

The response object is supposed to contain both `data` and `paginationInfo` keys

**Actual behaviour**

Only repos are returned.

**Comment by jdalrymple** [comment]
Weird, Ill give it a look and follow up!

**Comment by jdalrymple** [comment]
Testing this with the latest release without an issue. Is there any other information you could provide to aid me in narrowing down the problem?

**Comment by jdalrymple** [comment]
Note that your code needs to be run within a function, and that await cannot be used at the root level.

```js
async function test(){
  const projects = new Projects({token: xxxxx})
  
  const resp = await projects.all({
        showExpanded: true,
        perPage: 20,
        maxPages: 2,
        minAccessLevel: 40,
      })
  
  console.log(resp.data) // undefined
  console.log(resp.paginationInfo) // undefined
  console.log(resp) // array of repos 
}

test()
```

**Comment by jdalrymple** [comment]
Closing due to inactivity

**Comment by JounQin** [comment]
I have this issue too, and the ts type is unexpected.

![image](https://user-images.githubusercontent.com/8336744/129139114-b6174b07-1afc-49f0-8def-f86f66bf926c.png)

**Comment by JounQin** [comment]
@jdalrymple 

OK, I just found `showExpanded` is actually unavailable in ts options. So this should a bug.

**Comment by jdalrymple** [comment]
Ill give it a look! Thanks for the extra info

**Comment by jdalrymple** [comment]
Hmm, it looks like the typing is not being picked up when its that nested. Not sure why

**Comment by jdalrymple** [comment]
Damn, this is a bigger problem. Gonna need to dig deeper https://github.com/microsoft/TypeScript/issues/33369

**Comment by JounQin** [comment]
Function overload can be used for typings.

**Comment by jdalrymple** [comment]
haha easier said than done. Using overloads results in many many many more functions. Thats what lead me to use conditional types. Im going to spend a bit of time looking at what these limitations are and if there is a way to work around them.

**Comment by jdalrymple** [comment]
Noted: https://stackoverflow.com/questions/64650888/typescript-not-inferring-type-from-outer-function

**Comment by jdalrymple** [comment]
Testing

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzmAhlAzgUzi9y1YAywAdgNZwC+cAZlBCHAOQA2wTA3AFCiSyL4MmAKpQWOPKiEBFAK6YoATwDKMKKQDmVWvUZMAjvKUBadGs2ce4aPCQBjFCExsAXpgDSmRXmp0GzAA8HLEsuAHowuAAlTENMMzgAeTAYYAgSdC5MAL54GEUwbABJdAAxaEYAXkQuODhgMoqAfgAuOAAjCAgWTBQSbkpuLJybOHzCuGVZABMIOGqEWrh0GYhW5fMSLQAfOBJZEHaFAaHs3LGC7GUACwgAdwBRHL7pzGmAHgAVOGyYTBJpnhOt1eiR5h0uj0+gA+cGLOroW6PZ4At7rT4nLjDc7jbAAIVwmBicTMyVS6Tw1SmszgADJopg7NAPmZ1FsADRwWTkEj3EjQ04jfi4uAErAABRQGlIKHJJGJ8lJKTSGXe4p+AT+ALwTDIXiwMCYcF2TAgNBoBqN1V1+swhuNzDNFrtTFhlSWYqJsUVMDJKrw9PhdXw0pIspV63F3GDyAUko0mHW+0OxyWgyxZ1GIsS5oN8ZlcoV8V9yopcKWqATSYORyg0bgIBQAXj8WrKbrXHT2Kzlzg+bDf2mRaVcvQao1WsBzD13hdDtNubn1pnlvnTstbqW6t+-ynNtnhqWdSaosJ-fD6WHJdHauhR7gbU958L3uLfopt7pSUXMGfKqv74ZEMEQMugkAZNg5RQI2MCZJm-CkH8UA0CgdjYE8qCokO8TgVgXzgjETJQCymwaJy3JkLydz8rCQbTLKKBtBiSzXL0rwYG0hHMu8rKaORPJ8gKSxmLKsjoG0ya1gM3YISQSEoWhfZSgW-44RSmD4dUXHETxpH8ZRgkANoALq0Us9EwIxcDMXUlYqekRQkDQEBtEGdQwBAlksBJNapjGJC-D57YOvsLAsPWdR2LIUBQP8MBBVJ95gLFABuaRiQlCghbIYURbGUAtplHYxh5XktuJey+cVVDSTJeS9k8k5vNp2EwNFJDvEsDwTruQKQqC7JLN8O7agyREkWyZFcgJ1GwrsKUQMA0yDW61k9aNC1LUsJ6bdMSxtN1I1TihLBYNt1n7XAGEvM1akQV8Ql1RcEyNbubwAIIxSgigxG1UAdV161TsCUIkINdSfODfZAzqK5zia64uit4KHZqvW0Cgp2YOdnwmZd25o6N+6WjjeN1G0f6XndeG46Zgo4g1hOvK17WdXUqOTn1IJ9FDBOc9Otr2gjP5MFDw1M3gLW6ZN+lUfyDpS7xHLTQZs0mcj1Ti-zit6Srct3seV1M7d42-azDycp8BsPmtR14AAFKQNBZVEACUZNwCer1YZ9UDfWb-3vBb0ScuK1sSZgKWplwuCKCQdi0NydhynACYwAAEs4hRQOOdsCweRrC86hqchz6Mg6C0L28JChpWhnGMtxStTRR+tQ7ukCIW0zdQ25yyrFDCJItdWGDw2TblWPAB0M9xEoaaPmeykDqpJLXv6n70jc9wj8zQewoG4-NlKrZFdoCyUFDqF2A343S3xetGcZ4Lq1wrsU7oDQaX0ihmZFFLwFimBdS4IUB3BQMAeAWAoB10wFPWKa8FBTzTvbDui05KciQHPRQnIVg0koK7esTIMi2DgKxFA7F0C4Msm1Hw4IgG4UwPWHopDOjTEUOfOADD1JDDqCBAAwo4ZwwA3BcOptgNhHDgA0GWIUOw0jgBvCWNIuA9toGwKng4JwrhMCuwhOw8EWjhFuE8N4e2kjCFYj4ZEdOLwehwAgO0AAVoyQB4jMh1BUfbAAhL7b6U8Gh+MUOYiA7DXZ6L7l47xiId4omZno2Kf0wSSN4cGRJ7UagxjqBZKykix51HIZQ-JGxRJUPvOmOolArFwBArYgE9i0B+w4dwiCHi4DEISAFO4707AJ2qIZGeU9r6ckGZI4yRCAECACpqKgrk9iBQ2JNThUgiCkDIPbQpCh0BTzYOQSx-8SECGwW0fQlJEDUGoNUaZ8ATwrMwHIBQKhSL22uZgwQWAAByVUKpqHkFQPRrkKntMmXcSB1xSB4ggNyKc1RGzHwTG0w2XSel2B2f8DQMBrg1NUfbU5U9s4AH1KzYFwI-aiEhKrtj0dsXYAAmAADHo94R9yr3jaL8ph1SQKEWilgMY1x6CyA0FiuyK90hiJWCwWCyiZE+NxUYRQ+KT50npNfNFWxMXzEqNURlKr5kzNpPSUFmKIVQu1BE+86T-qpztJnFg2dq5ZOWLXYAaFiloMQsUvuMZBmnOKXUOFk97wxjwRAYpl9g3zO6b0se+yqDVKiTE5EmFmbGl2NgpVoYLxgm1cuQWTAEl2gyci3pqSrVgj7jkiO0a7Bj1FdmxyzlXKRtKpjNodzHIwA2WxLZhkmABGMK2lgTBjKcgAIyMuKdc9tBBMCdu7RQ3t-bjDXOMMSkd47dU0sqmFYpUUYpxRnUIedmyMB9oHeu0dcAJ3Ut2GO4pyVI7pQqh2uSC7KHnrXalNdJ8N3Xq3bsUKLAH1xhPkerAJ6e1nuXdnH9CY-03uKUO8q4G51vtPegT9Q64PxAQ5O8pUN0xVKejQJOKc05s1thLMaTddat0EgrRuOke5kv5CZAiTGJoP3o7NFa9sLV1HLXAXO1Hibw0dCLcEC5i6iyujDCE3Mc0YyxlXe86jXWYFvrRmWrHoRjw9XJbuel7wQFLBkSMy9ZS3TXoBMcYcvzb2TTdD4DwD40eY7rH+emljvz7J-PC3tmYBw6sHcUltoRulhGnO1Dr1Nup+ACTuGCHFmfQJY4jERPgPGUJ8IoHyADihWivFZK6Vsr5WKuVaK1wDp8A+k2q7a7e2CBL7MFk0gJNu83jsqgH8ghWIuCkfjinWU-HMlpKLdalBTWWuciYO15Yw84ndbGL17A-XiNDeTiqWgEB7ZrDmZ15b0x1gVz6AQ8bgnJtgmm811r83OQQFdp2LEtWuHgmcs1o7KaVscoIUAA

**Comment by jdalrymple** [comment]
It looks like the only solution (besides writing many many many overload functions) is to add generic variables to each function like:

```
function fnname<P extends 'keyset' | 'offset' = 'offset', E extends boolean = false>(o?: ShowExpanded<E> & PaginatedRequestOptions<P>)
```

Kind of annoying, but until i can think of a better way fo doing it, ill implement this where needed

**Comment by henrysachs** [comment]
not to put pressure onto this but whats your status on this? would be highly appreciated because i could need the information provided by showExpanded. :)

**Comment by jdalrymple** [comment]
This is the next thing ill be jumping on! Just been swamped lately. I did try to implement what i came up with earlier (see previous comment), but i ran into an issue when actually putting it into practice. I should have some free time by the end of this week, ill follow up then :pray:

**Comment by jdalrymple** [comment]
:rocket: Issue was released in [`37.0.0`](https://github.com/jdalrymple/gitbeaker/releases/tag/37.0.0) :rocket:
